cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -308,11 +309,24 @@ class DatasetWriterDirectoryQueue {
   }
 
   Result<std::string> GetNextFilename() {
-    auto basename = ::arrow::internal::Replace(write_options_.basename_template,
-                                               kIntegerToken, ToChars(file_counter_++));
+    std::optional<std::string> basename;
+    if (write_options_.basename_template_functor == nullptr) {
+      basename = ::arrow::internal::Replace(write_options_.basename_template,
+                                            kIntegerToken, ToChars(file_counter_++));
+    } else {
+      basename = ::arrow::internal::Replace(
+          write_options_.basename_template, kIntegerToken,
+          write_options_.basename_template_functor(file_counter_++));
+    }
     if (!basename) {
       return Status::Invalid("string interpolation of basename template failed");
     }
+    if (used_filenames_.find(*basename) != used_filenames_.end()) {
+      return Status::Invalid("filename ", *basename,
+                             " is already used before. Check basename_template_functor");
+    } else {
+      used_filenames_.insert(*basename);
+    }
```

Comment:
Small nit, not a big deal, but you could combine the `find` and `insert` calls...

```
if (!used_filenames_.insert(*basename).second) {
  return Status::Invalid(...);
}
```